src/utils/user.authentication.ts

```diff
@@ -10,11 +10,20 @@ import {promisify} from 'util';
 import * as isemail from 'isemail';
 import {HttpErrors} from '@loopback/rest';
 import {UserProfile} from '@loopback/authentication';
-import {compare} from 'bcryptjs';
-const compareAsync = promisify(compare);
+import {compareSync, genSaltSync} from 'bcryptjs';
```

Comment:
Is it a good idea to use sync functions? I think this will block the event loop for non-negligible amount of time and can seriously degrade application's performance.

Can we use `compareAsync` instead? Same for `genSalt`.

Reply:
@bajtos I got a problem when using the promisified `genSalt` with signature `genSaltAsync(10)`, opened a PR in `@types`, you can check the detail in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32868

That's the reason why I switched to `genSaltSync`, and I thought it will save the effort of doing promisify, therefore switched to sync styles for other functions.

> I think this will block the event loop for non-negligible amount of time and can seriously degrade application's performance.

But I now feel your point makes more sense. The time caused by blocking the event loop is more than the promisify time.

Reply:
> I got a problem when using the promisified `genSalt` with signature `genSaltAsync(10)`, opened a PR in `@types`, you can check the detail in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32868
>
> That's the reason why I switched to `genSaltSync`, and I thought it will save the effort of doing promisify, therefore switched to sync styles for other functions.

Until the typings are fixed, you can use an explicit type cast as a workaround. Just leave a comment explaining why the workaround is needed + a link to the GH issue you created.

For example:

```ts
const genSaltAsync = promisify(genSalt) as (rounds: number) => Promise<string>;
// or disable argument/retval type checks 
const genSaltAsync = promisify(genSalt) as Function;
```

We should get the implementation right and use async versions. If we need to bend the type system to do so, then let's do it.